File: aalto/week2/monday/escape.py
def bfs(nodes_dict, moped, police, intersections):
    for i in range(intersections):
        if i not in nodes_dict.keys():
            print("-1")
            continue
        depth = 0
        fifo = [i]
        discovered = []
        escape = False
        while fifo:
            n = fifo.pop(0)
            discovered.append(n)
            for node in nodes_dict[n]:
                if node[0] not in discovered and node[1] >= moped:
                    if node[1] >= moped and node[1] < police:
                        escape = True
                        depth += 1
                        break
                    fifo.append(node[0])
            if escape: 
                print(depth)
                break
            depth += 1
        if not escape: print("-1")

def read():
    first = input().split(" ")
    intersections = int(first[0])
    roads = int(first[1])
    moped = int(first[2])
    police = int(first[3])
    nodes_dict = {}
    for i in range(roads):
        road = input().split(" ")
        n1 = int(road[0])
        n2 = int(road[1])
        width = int(road[2])
        if n1 in nodes_dict.keys(): nodes_dict[n1].append((n2, width))
        else: nodes_dict[n1] = [(n2, width)]

        if n2 in nodes_dict.keys(): nodes_dict[n2].append((n1, width))
        else: nodes_dict[n2] = [(n1, width)]
    return nodes_dict, moped, police, intersections



def main():
    nodes_dict, moped, police, intersections = read()
    bfs(nodes_dict, moped, police, intersections)
    
main()

File: aalto/week2/monday/escape.cpp
#include <iostream>
#include <vector>
#include <array>
#include <unordered_map>
#include <queue>
#include <algorithm>
#define VI vector<int>
#define VAI vector<array<int, 2>>
#define MPVI unordered_map<int, VAI>


using namespace std;

void read(int* n, int* m, int* x, int* y, MPVI* nodes){
    cin >> *n >> *m >> *x >> *y;
    for (int i = 0; i < *m; i++)
    {
        int a, b, w;
        cin >> a >> b >> w;
        (*nodes)[a].push_back({b, w});
        (*nodes)[b].push_back({a, w});
    }
}

void bfs(int n, int m, int x, int y, MPVI nodes){
    for (int i = 1; i <= n; i++)
    {
        int depth = 0;
        queue<int> fifo;
        fifo.push(i);
        VI discovered;
        bool escape = false;
        while (!fifo.empty()){
            int c = fifo.front();
            fifo.pop();
            discovered.push_back(c);
            for (int j = 0; j < nodes[c].size(); j++)
            {
                int neighbour = nodes[c][j][0];
                int width = nodes[c][j][1];
                vector<int>::iterator it;
                it = find (discovered.begin(), discovered.end(), neighbour);
                if (it != discovered.end() || width < x) continue;
                if (width < y){
                    escape = true;
                    break;
                }
                fifo.push(neighbour);
            }
            depth++;
            if (escape) break;
        }
        if (escape) cout << depth << endl;
        else cout << "-1" << endl;
    }
    
}

int main(){
    int n, m, x, y;
    MPVI nodes;
    read(&n, &m, &x, &y, &nodes);
    bfs(n, m, x, y, nodes);
    return 0;
}